[Federation] Add type-agnostic e2e crud test

marun opened the pull request:
This PR proposes an e2e test that reuses the type-agnostic crudtester already used for integration testing to validate crud against a deployed cluster. It is intended to to eventually replace the existing e2e tests for simple types like secrets, but for now will run in addition to the existing testing to gain confidence in the coverage it provides.

The deletion corner cases - when orphanDependents is nil or true - do not involve operations in the member clusters and are already well-tested in integration testing and so are not reimplemented here.

Where it can be applied, this approach of abstracting a test from its execution environment - making the test 'retargetable' - can reduce the cost of test development since the bulk of the work can be iterated on as an integration test.  It can also serve as a check on assumptions made in the integration test about how a deployed environment will behave.

cc: @kubernetes/sig-federation-pr-reviews @kubernetes/sig-testing-misc @smarterclayton @derekwaynecarr

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43194)
<!-- Reviewable:end -->

madhusudancs commented:
@nikhiljindal assigning it to you because you are reviewing other integration test PRs and have better context about it.

marun commented:
cc: @perotinus

perotinus commented on test/integration/federation/framework/secret.go:
```diff
@@ -31,7 +31,7 @@ import (
 
 type SecretAdapter struct {
-	client federationclientset.Interface
+	Client federationclientset.Interface
```
Does this need to be public? You could add a NewSecretAdapter method instead.

marun replied:
Is there a reason to choose one approach over another?  I've generally preferred to use factory methods only when more than simple assignment is required.

perotinus reviewed (commented):
This looks good overall. I do question whether the CRUDHelper should be in a separate package that is depended on by both integration and e2e tests: it seems more like a generic testing harness than something directly tied to either kind of test.

marun commented:
@perotinus I agree that CRUDHelper should be separated, if only because otherwise any e2e using it will be pulling in everything needed to run federation and kubernetes api servers.  I think having it live in something like ``test/integration/federation/crud`` could be ok, though.  I don't think there is precedent for sharing test code between integration and e2e, and I'm not sure it makes sense to specialize its placement just yet.

perotinus commented:
@marun Yes, that's kind of what I had in mind too. I think your idea of package structure is reasonable for now. I'm not familiar enough with the test tree to really know of a better place, though I could see either a federation subdir `fixtures` or `utils` being appropriate.

This LGTM.

marun committed:
fed: Add type-agnostic e2e crud test

marun commented:
I've rebased this on top of #43500 to benefit from the refactoring it did, and switched to adding a new test instead of replacing the old one.  My intent is to reduce the risk of losing coverage while the kinks in this approach are worked out.

perotinus commented on test/e2e_federation/crud.go:
```diff
@@ -0,0 +1,52 @@
+			It("should be created, read, updated and deleted successfully", func() {
+				fedframework.SkipUnlessFederated(f.ClientSet)
+
+				// Only need to load clients once
```
Why do you need to do this in the test block? Will GetClusterClients not work in the outer scope?

marun replied:
Getting the cluster clients is an expensive operation.  No point in doing it if the test will be skipped.

perotinus commented:
LGTM.

perotinus commented:
@csbell @nikhiljindal @madhusudancs Can someone else approve this?

madhusudancs commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented:
@k8s-bot bazel test this

marun commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented:
rebased

@k8s-bot pull-kubernetes-federation-e2e-gce test this

marun commented:
@csbell e2e is green: http://prow.k8s.io/?type=presubmit&pull=43194&job=pull-kubernetes-federation-e2e-gce

csbell commented:
/lgtm
